blaze/expr/collections.py

```diff
@@ -418,6 +418,7 @@ def on_right(self):
         return on_right
 
     @property
+    @memoize
     def schema(self):
```

Comment:
```python
class Join:
    ...
    class schema:
        def __get__(self, instance, owner):
            if instance is None:
                return self
            ret = <compute schema>
            instance.schema = schema
            return ret
    schema = schema()
```
becase we don't have a `__set__` or `__delete__` on `schema`, it is a non-data descriptor and once we have assigned a value to the instance, that will be looked up first. The class does not need to be defined inline of course, we could generalize this into some `lazyval` class.